fix: incoherency allowed characters tableName/columnName (+ improved regex)

svandenhoek opened the pull request:
What are the main changes you did:
- Fixed incoherency for what is allowed for tableName/columnName
- Improved regex: https://regex101.com/r/DD7c6W/1


how to test:
- explain here what to do to test this (or point to unit tests)

todo:
- [ ] updated docs in case of new feature
- [ ] added/updated tests
- [ ] added/updated testplan to include a test for this fix, including ref to bug using # notation

mswertz reviewed (commented):
looks good, let me know when finished

svandenhoek commented:
integrated #4172

mswertz commented on apps/molgenis-components/src/components/constants.ts:
```diff
@@ -3,6 +3,8 @@ export default {
   CODE_PERIOD: 46,
   CODE_MINUS: 45,
+  TABLE_NAME_REGEX: /^(?!.* _|.*_ )[a-zA-Z][a-zA-Z0-9 _]{0,30}$/,
+  COLUMN_NAME_REGEX: /^(?!.* _|.*_ )[a-zA-Z][a-zA-Z0-9 _]*$/,
```
I think there is a maximum length here as well, at least dictated by postgresql (maybe 64?)

svandenhoek replied:
It seems to be 63 (`length NAMEDATALEN - 1`) [1,2,3]. Will change accordingly.


Sources:
[1] https://www.postgresql.org/docs/current/runtime-config-preset.html#GUC-MAX-IDENTIFIER-LENGTH
[2] https://www.postgresql.org/docs/current/sql-syntax-lexical.html#SQL-SYNTAX-IDENTIFIERS
[3] running `psql -U molgenis -c "SHOW max_identifier_length;"` locally

mswertz commented on docs/molgenis/use_schema.md:
```diff
@@ -23,14 +23,21 @@ You can describe basic columns using:
 ### tableName
 
-Will be the name of the table. Must start with one of a-zAZ followed by zero or more of \_a-zAZ1-3. Maximum length 31 characters. If you leave columnName empty
+Regular expression requirement: `^(?!.* _|.*_ )[a-zA-Z][a-zA-Z0-9 _]{0,30}$`
```
I would like first the human readable and then the regexp.

mswertz reviewed (changes requested):
nice. Small comments. I particular make sure that user documentation starts with human explanation and then add regexp for the tech reader.

davidruvolo51 commented on e2e/tests/central/admin!create-database.spec.ts:
```diff
@@ -0,0 +1,30 @@
+import { test, expect } from '@playwright/test';
+
+let regexErrorMessage='Table name must start with a letter, followed by letters/underscores/spaces/numbers (though no underscore preceded/followed by a space) and with a maximum of 31 characters, i.e. ^(?!.* _|.*_ )[a-zA-Z][a-zA-Z0-9 _]{0,30}$'
```
Unless the message is reassigned, `const` should be used here instead of `let`.

davidruvolo51 commented on apps/central/src/components/SchemaCreateModal.vue:
```diff
@@ -178,7 +179,7 @@ export default {
         return undefined;
       } else {
-        return "Table name must start with a letter, followed by letters, underscores, a space or numbers, i.e. [a-zA-Z][a-zA-Z0-9_]*. Maximum length: 31 characters";
+        return "Table name must start with a letter, followed by letters/underscores/spaces/numbers (though no underscore preceded/followed by a space) and with a maximum of 31 characters, i.e. ^(?!.* _|.*_ )[a-zA-Z][a-zA-Z0-9 _]{0,30}$";
```
Should we put the regex pattern in the error message? It is helpful for developers, but I'm not sure if many end users (data managers, lab personnel, etc.) would understand this.

svandenhoek replied:
Fair point. I left it in as they were present already, though for a user (without IT background) it indeed can just cause confusion.

davidruvolo51 commented on apps/schema/src/components/ColumnEditModal.vue:
```diff
@@ -445,8 +446,8 @@ export default {
-      if (!this.column.name.match(/^[a-zA-Z][a-zA-Z0-9_ ]+$/)) {
-        return "Name should start with letter, followed by letter, number, whitespace or underscore ([a-zA-Z][a-zA-Z0-9_ ]*)";
+      if (!this.column.name.match(constants.COLUMN_NAME_REGEX)) {
+        return "Name should start with a letter, followed by letters/underscores/spaces/numbers (though no underscore preceded/followed by a space), i.e. ^(?!.* _|.*_ )[a-zA-Z][a-zA-Z0-9 _]*$";
```
(See comment above)

davidruvolo51 commented on apps/schema/src/components/TableEditModal.vue:
```diff
@@ -143,9 +144,9 @@ export default {
+        this.table.name.search(constants.TABLE_NAME_REGEX)
       ) {
-        return "Name is required and can only contain 'azAZ_ '";
+        return "Name is required and must start with a letter, followed by letters/underscores/spaces/numbers (though no underscore preceded/followed by a space) and with a maximum of 31 characters, i.e. ^(?!.* _|.*_ )[a-zA-Z][a-zA-Z0-9 _]{0,30}$";
```
(Same as above)

davidruvolo51 reviewed (changes requested):
Looks good. I think some minor revisions to the error message could help make it clearer for end users. Perhaps something like this:

> Table names must start with a letter followed by one or more letters, numbers or underscores. Names cannot exceed 31 characters. Spaces are also allowed as long as they do not occur immediately before or after an underscore. 

We could also have a few examples: `My table`, `myTable`, `my_table`, `My table 2`, etc.

svandenhoek commented:
@mswertz:
* Regex now after human-readable in docs.
* Added limit (`63`) to column names.

@davidruvolo51:
* A table with examples is added for database creation docs and table/column names refer to that table for examples.
* Renamed error messages (without regex, though slightly deviated from your example).
* Fixed `let` -> `const`

davidruvolo51 reviewed (approved):
Looks great!
